src/main/java/org/gradle/profiler/maven/MavenScenarioInvoker.java

```diff
@@ -1,25 +1,53 @@
 package org.gradle.profiler.maven;
 
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import org.gradle.profiler.BuildToolCommandLineInvoker;
-import org.gradle.profiler.GradleBuildConfiguration;
+import org.gradle.profiler.InstrumentingProfiler;
+import org.gradle.profiler.InstrumentingProfiler.SnapshotCapturingProfilerController;
 import org.gradle.profiler.InvocationSettings;
-import org.gradle.profiler.Logging;
+import org.gradle.profiler.ScenarioSettings;
 import org.gradle.profiler.result.BuildInvocationResult;
 
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.function.Consumer;
 
 public class MavenScenarioInvoker extends BuildToolCommandLineInvoker<MavenScenarioDefinition, BuildInvocationResult> {
     @Override
-    public void run(MavenScenarioDefinition scenario, InvocationSettings settings, Consumer<BuildInvocationResult> resultConsumer) {
+    public void run(MavenScenarioDefinition scenario, InvocationSettings settings, Consumer<BuildInvocationResult> resultConsumer) throws IOException, InterruptedException {
         List<String> commandLine = new ArrayList<>();
         commandLine.add(scenario.getExecutablePath());
         commandLine.addAll(scenario.getTargets());
         scenario.getSystemProperties().forEach((key, value) ->
             commandLine.add(String.format("-D%s=%s", key, value)));
 
-        doRun(scenario, settings, resultConsumer, commandLine);
+        Map<String, String> profileEnvironment;
+        SnapshotCapturingProfilerController controller;
+        // TODO This only works with Async profiler, since the only thing we call from the controller is stopSession()
+        //      Capture this in the type hierarchy somehow
+        if (settings.getProfiler() instanceof InstrumentingProfiler) {
+            InstrumentingProfiler profiler = (InstrumentingProfiler) settings.getProfiler();
```

Comment:
🤔 Do we need this cast?

Reply:
Oh I guess because of profiler.newSnapshottingController`, I wonder how we could make that nice 🤔

Reply:
This is what I dislike about this, yes. I think we should re-model how profilers and build tools work together, but that's outside the scope of this PR.

Reply:
Let's continue on https://github.com/gradle/gradle-profiler/pull/576